astropy/coordinates/matching.py

```diff
@@ -250,8 +251,12 @@ def search_around_3d(coords1, coords2, distlimit, storekdtree="kdtree_3d"):
     considered an implementation detail, though, so it could change in a future
     release.
     """
-    if not distlimit.isscalar:
-        raise ValueError("distlimit must be a scalar in search_around_3d")
+    if not distlimit.isscalar and len(distlimit) != len(coords1):
```

Comment:
I don't think it is necessary to raise a special error here - just write the code such that it will error (would be `strict=True` in your `zip`, but my `np.broadcast_arrays` will do it for free).